src/MongoDB.Analyzer/Core/Builders/BuildersResolveVariables.cs

```diff
@@ -0,0 +1,378 @@
+        
+        if (canEvaluate)
+        {
+            buildersNodes.Add(syntaxNode);
+        }
+        else if (syntaxNode is BinaryExpressionSyntax binaryExpression)
+        {
+            canEvaluate = CanEvaluate(binaryExpression.Left, buildersNodes, contextInformation) &&
+                          CanEvaluate(binaryExpression.Right, buildersNodes, contextInformation);
+        }
+
+        return canEvaluate;
+    }
+
+    private static (SyntaxNode RewrittenBuildersExpression, List<ConstantsMapper> constantsMappers, string ArgumentTypeName) RewriteBuildersExpression(
+        SyntaxNode syntaxNode,
+        List<SyntaxNode> childBuilderNodes,
+        ContextInformation contextInformation)
+    {
+        var syntaxKind = syntaxNode.Kind();
+        var argumentTypeName = "";
+        var constantMappers = new List<ConstantsMapper>();
+        var nodesRemapping = new Dictionary<SyntaxNode, SyntaxNode>();
+        var variableValues = contextInformation.VariableValues;
+        var builderToAnalysisContext = contextInformation.BuilderToAnalysisContext;
+
+        foreach (var builderNode in childBuilderNodes)
+        {
+            SyntaxNode rewrittenBuildersNode = null;
+            ConstantsMapper constantsMapper = new ConstantsMapper();
+
+            if (builderNode is IdentifierNameSyntax identifierName)
+            {
+                var variableValue = variableValues[identifierName.Identifier.ValueText];
+                var analysisContext = variableValue.ExpressionAnalysisContext;
+                rewrittenBuildersNode = analysisContext.Node.RewrittenExpression;
+                constantsMapper = analysisContext.Node.ConstantsRemapper;
+                argumentTypeName = analysisContext.Node.ArgumentTypeName;
+                variableValue.Locations.Add(builderNode.GetLocation());
+            }
+            else
+            {
+                var analysisContext = builderToAnalysisContext[builderNode];
+                rewrittenBuildersNode = analysisContext.Node.RewrittenExpression;
+                constantsMapper = analysisContext.Node.ConstantsRemapper;
+                argumentTypeName = analysisContext.Node.ArgumentTypeName;
+            }
+
+            nodesRemapping.Add(builderNode, rewrittenBuildersNode);
+            constantMappers.Add(constantsMapper);
+        }
+
+        var rewrittenBuildersExpression = syntaxNode.ReplaceNodes(nodesRemapping.Keys, (n, _) => nodesRemapping[n]);
+
+        if (syntaxKind == SyntaxKind.AndAssignmentExpression ||
+            syntaxKind == SyntaxKind.OrAssignmentExpression)
+        {
+            var assignmentExpression = rewrittenBuildersExpression as AssignmentExpressionSyntax;
+            var identifierSyntaxNode = assignmentExpression.Left as IdentifierNameSyntax;
+            var identifier = identifierSyntaxNode.Identifier.ValueText;
+            VariableValue identifierValue = null;
+            variableValues.TryGetValueSafe(identifier, out identifierValue);
```

Comment:
```
if (!variableValues.TryGetValueSafe
  return default;
```
but that should never happen?

Reply:
This can happen in cases for y & = x when "y" is removed (i.e. due to different context)